### PR TITLE
docs: Document SpinLock and RWLock

### PR DESCRIPTION
**Pull request by Zildj1an**
Document locking support. This is related to issue #74 Document COCONUT-SVSM.

### CODE REVIEW

**Review by 00xc — COMMENTED** [review]
Looks good, just a few notes:

* When mentioning a type (e.g. `RWLock`, `WriteLockGuard`) I'd add a link to the type with square brackets.
* I'd avoid documenting trait implementations unless something counterintuitive is going on, as the docs for the trait itself should suffice. The generated docs already include links to the Rust core library documentation.
* This is just a preference, but I'd hide imports in doctests with `#` to clean the generated docs.

**Comment by joergroedel** [comment]
> Looks good, just a few notes:
> 
> * When mentioning a type (e.g. `RWLock`, `WriteLockGuard`) I'd add a link to the type with square brackets.
> * I'd avoid documenting trait implementations unless something counterintuitive is going on, as the docs for the trait itself should suffice. The generated docs already include links to the Rust core library documentation.
> * This is just a preference, but I'd hide imports in doctests with `#` to clean the generated docs.

Given these discussions it would be helpful if you two can come up with a documentation style document. For the virtual memory code the doc domments look quite different and some common guidelines will help to better align the style.

**Comment by Zildj1an** [comment]
I have applied your feedback @00xc -- except for hiding imports, because I  personally prefer seeing them; we can discuss this while preparing guidelines as @joergroedel suggested.

File: src/locking/spinlock.rs [hunk 1]
```diff
@@ -9,43 +9,107 @@ use core::fmt::Debug;
 use core::ops::{Deref, DerefMut};
 use core::sync::atomic::{AtomicU64, Ordering};
 
+/// A lock guard used in conjunction with [`SpinLock`]. This lock guard
```
Comment (00xc) [inline src/locking/spinlock.rs:12]: I'd make it explicit that one obtains this type from a `SpinLock`:

```suggestion
/// A lock guard obtained from a [`SpinLock`]. This lock guard
```

File: src/locking/spinlock.rs [hunk 2]
```diff
@@ -9,43 +9,107 @@ use core::fmt::Debug;
...
+/// provides exclusive access to the data protected by a [`SpinLock`],
+/// ensuring that the lock is released when it goes out of scope.
+///
+/// # Examples
+///
+/// ```
+/// use svsm::locking::{SpinLock, LockGuard};
```
Comment (00xc) [inline src/locking/spinlock.rs:19]:
```suggestion
/// use svsm::locking::SpinLock;
```

File: src/locking/spinlock.rs [hunk 3]
```diff
@@ -9,43 +9,107 @@ use core::fmt::Debug;
...
+    /// This atomic counter is incremented each time a thread attempts to
+    /// acquire the lock. It helps to determine the order in which threads
+    /// acquire the lock.
     current: AtomicU64,
+    /// This counter represents the thread that currently holds the lock
+    /// and has access to the protected data.
     holder: AtomicU64,
+    /// This `UnsafeCell` is used to provide interior mutability of the
+    /// protected data, this is, it allows the data to be accessed/modified
+    /// while respecting the locking mechanism.
```
Comment (00xc) [inline src/locking/spinlock.rs:95]:
```suggestion
    /// protected data. That is, it allows the data to be accessed/modified
    /// while enforcing the locking mechanism.
```

**Comment by Zildj1an** [comment]
Thanks @00xc I applied the feedback

**Review by 00xc — APPROVED** [review]

**Pull request merged**